src/client/src/producer.rs

```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+                    partition
+                }
+            };
+            partitions.push(partition);
+        }
+        *self.index.write().expect("poisoned") = rr_index;
+        partitions
     }
 }
 
-/// Sends record to a target server (Kf, SPU, or SC)
-async fn send_records_raw<F: SerialFrame>(
-    mut leader: F,
-    replica: &ReplicaKey,
-    records: Vec<(Option<Bytes>, Bytes)>,
-) -> Result<(), FluvioError> {
-    use dataplane::produce::DefaultProduceRequest;
-    use dataplane::produce::DefaultPartitionRequest;
-    use dataplane::produce::DefaultTopicRequest;
-    use dataplane::batch::DefaultBatch;
-    use dataplane::record::DefaultRecord;
-    use dataplane::record::DefaultAsyncBuffer;
-
-    // build produce log request message
-    let mut request = DefaultProduceRequest::default();
-    let mut topic_request = DefaultTopicRequest::default();
-    let mut partition_request = DefaultPartitionRequest::default();
-
-    debug!("Putting together batch with {} records", records.len());
-
-    let records = records
-        .into_iter()
-        .map(|(key, value)| {
-            let key = key.map(DefaultAsyncBuffer::new);
-            let value = DefaultAsyncBuffer::new(value);
-            DefaultRecord::from((key, value))
-        })
-        .collect();
-
-    let batch = DefaultBatch::new(records);
-    partition_request.partition_index = replica.partition;
-    partition_request.records.batches.push(batch);
-    topic_request.name = replica.topic.to_owned();
-    topic_request.partitions.push(partition_request);
-
-    request.acks = 1;
-    request.timeout_ms = 1500;
-    request.topics.push(topic_request);
-
-    trace!("produce request: {:#?}", request);
-
-    let response = leader.send_receive(request).await?;
-
-    trace!("received response: {:?}", response);
-
-    // process response
-    match response.find_partition_response(&replica.topic, replica.partition) {
-        Some(partition_response) => {
-            if partition_response.error_code.is_error() {
-                return Err(IoError::new(
-                    ErrorKind::Other,
-                    partition_response.error_code.to_sentence(),
-                )
-                .into());
+fn partition_siphash(key: &[u8], partition_count: i32) -> i32 {
+    use std::hash::{Hash, Hasher};
+    use std::convert::TryFrom;
+
+    assert!(partition_count >= 0, "Partition must not be less than zero");
+    let mut hasher = SipHasher::new();
+    key.hash(&mut hasher);
+    let hashed = hasher.finish();
+
+    i32::try_from(hashed % partition_count as u64).unwrap()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_u64_mod_i32() {
+        assert_eq!(i32::MAX as u64, 2147483647_u64);
+        let some_u64 = 2u64.pow(63);
+        let some_u64_mod_i32 = some_u64 % (i32::MAX as u64);
+        assert!(some_u64_mod_i32 < i32::MAX as u64);
+        let some_i32 = some_u64_mod_i32 as i32;
+        println!(
```

Comment:
prefer not to have println in test code.  if need to diagnose internal state, put as debug or should do another assert